# neuralmt: default program

In [19]:
from neuralmt import *
import os, sys
import sacrebleu

def bleu(ref_t, pred_t):
    return sacrebleu.corpus_bleu(pred_t, [ref_t], force=True, lowercase=True, tokenize='none')

## Run the default solution on dev

In [10]:
model = Seq2Seq(build=False)
model.load(os.path.join('..','data', 'seq2seq_E049.pt'))
#device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device = torch.device('mps' if torch.backends.mps.is_built() else 'cpu')
model.to(device)
model.eval()
# loading test dataset
test_iter = loadTestData(os.path.join('..','data', 'input', 'dev.txt'), model.params['srcLex'],
                            device=device, linesToLoad=sys.maxsize)
results = translate(model, test_iter) # Warning: will take >5mins depending on your machine
print("\n".join(results))

1305it [05:01,  4.32it/s]

when i was in my , i got my first . 
i was a and studied clinical psychology . 
she was a female woman . 
when i came back first , she had her and and a a . she took in the office office , her her , and her told me she was to to talk about about . 
and i heard that . i was . 
my first business was a a . 
and i got a woman in the who who wanted to talk . 
i i i . 
but i did . 
with the stories , the alex with the , it was easy easy to me simply with the head , while we the problems problems . 
" 30 is the new 20 " , and i i was able to 
work later , later later later later later later later later death later later later . 
people in the and i did nothing as a time . 
but soon , my my , , , to my . 
i hold . 
i said , " yeah , she 's with men with her her , she with with a , but she wo n't marry him . " 
and i said , " not yet , but better better . 
also , the best time to say is marriage is that before they married . " 
that 's what psychologists call a . 
that was the moment that i rea

## Evaluate the our output

In [22]:
#from bleu_check import bleu
ref_t = []
with open(os.path.join('..','data','reference','dev.out')) as r:
    ref_t = r.read().strip().splitlines()
print(bleu(ref_t, results))

BLEU = 14.24 60.3/26.0/12.7/6.6 (BP = 0.749 ratio = 0.775 hyp_len = 19310 ref_len = 24902)


## Documentation

Write some beautiful documentation of your program here.

### Implementation
#### Remove adjacent duplicate words
```
def remove_adjacent_duplicates(text):
    """
    Removes adjacent duplicate words from the given text.

    Parameters:
    - text (str): The input text containing words.

    Returns:
    str: The modified text with adjacent duplicate words removed.
    """
```
#### Attention model
`def calcAlpha(self, decoder_hidden, encoder_out)`: Calculate attention score

`def forward(self, decoder_hidden, encoder_out)`: Implement the forward function for attention model

## Analysis

Do some analysis of the results. What ideas did you try? What worked and what did not?

### Worked
* Attention model:
    * BLEU score has been improved to 14.24.
    * There are significant improvements in translation quality.
    * Richer content and fewer word repetitions

### Did not work:
* Remove adjacent duplicate words
    * the output looks better
    * the BLEU score has dropped, probably because of the way BLEU is calculated.
